x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/profile/SuggestProfilesRequest.java

```diff
@@ -66,6 +82,95 @@ public ActionRequestValidationException validate() {
         if (size < 0) {
             validationException = addValidationError("[size] parameter cannot be negative but was [" + size + "]", validationException);
         }
+        if (hint != null) {
+            validationException = validateHint(validationException);
+        }
+        return validationException;
+    }
+
+    private ActionRequestValidationException validateHint(ActionRequestValidationException validationException) {
+        final List<String> uids = hint.getUids();
+        final Map<String, List<String>> labels = hint.getLabels();
+
+        if (uids == null && labels == null) {
+            return addValidationError("[hint] parameter cannot be empty", validationException);
+        }
+
+        if (uids != null && uids.isEmpty()) {
+            validationException = addValidationError("[uids] hint cannot be empty", validationException);
+        }
+
+        if (labels != null) {
+            if (labels.size() != 1) {
+                return addValidationError(
+                    "[labels] hint supports a single key, got [" + Strings.collectionToCommaDelimitedString(labels.keySet()) + "]",
+                    validationException
+                );
+            }
+
+            final List<String> values = labels.values().iterator().next();
+            if (values.isEmpty()) {
+                return addValidationError("[labels] hint cannot be empty", validationException);
+            }
+        }
         return validationException;
     }
+
+    public static class Hint implements Writeable {
+        @Nullable
+        private final List<String> uids;
+        @Nullable
+        private final Map<String, List<String>> labels;
```

Comment:
The single key for the `labels` field is an intentional restriction that we agreed to keep the scope as minimal as possible to cater for only known, concrete requirements. We might decide to allow more keys in the future. But it's hard to tell for now. There is also inherent limitation on the expressiveness of a map structure. For example, we often interpret the relationship between multiple keys to be "AND". So there is no easy way to express "OR". 

> If that's not the case, might go with a record Label with name and values fields, or possibly including those fields on the Hint class directly. That way the structure is more explicit and validation and accessing labels gets simpler.

For just a single key, this makes sense. I considered that but didn't want to let the "design level restriction" impact on the underlying data struture. I have no strong opinion. If Albert also thinks it's better to get rid of the Map, I'll be happy to follow the suggestion.

> If we want to support multiple labels in the future, wondering what that will look like in terms of bwc
In this particular case, bwc is not a practical issue because the request should never be sent across nodes (also transport client is no longer a thing in 8.x). But if we do encouter such situation where the remote node cannot support what the local node requires, the common strategy is to throw error.

Reply:
Thanks for the context 👍 

>> If Albert also thinks it's better to get rid of the Map, I'll be happy to follow the suggestion.

Sounds good! I don't feel strongly either way, esp. if since based on what you said there _is_ a chance we will extend this to necessitate a map in the future

Reply:
> If Albert also thinks it's better to get rid of the Map, I'll be happy to follow the suggestion.

I think it's OK like it is.